torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1441,8 +1442,184 @@ void InitXlaModuleBindings(py::module m) {
+        });
+
+  // Return true if value of the tensor requires a computation.
+  m.def("_check_tensor_need_materialization",
+        [](const std::vector<at::Tensor>& tensors) -> std::vector<bool> {
+          std::vector<bool> need_materialization;
+          need_materialization.reserve(tensors.size());
+          for (auto& tensor : tensors) {
+            auto xtensor = bridge::TryGetXlaTensor(tensor);
+            if (!xtensor) {
+              // input tensor is not a xla tensor
+              need_materialization.push_back(false);
+            } else if (xtensor->CurrentXlaData() != nullptr) {
+              // input tensor has xla_data which means it is already on device
+              need_materialization.push_back(true);
+            } else if (xtensor->CurrentIrValue().node != nullptr) {
+              torch::lazy::NodePtr node = xtensor->CurrentIrValue().node;
+              if (torch_xla::DeviceData::Cast(
+                      xtensor->CurrentIrValue().node.get()) != nullptr) {
+                need_materialization.push_back(false);
+              } else {
+                // input tensor is an IR other than DeviceData which means a
+                // compuation is required to get the value of this tensor.
+                need_materialization.push_back(true);
+              }
+            } else {
+              // TODO: maybe also handle it is a XLATensor with tensor_data case
+              XLA_CHECK(false);
+            }
+          }
+          return need_materialization;
+        });
+
+  m.def("_get_graph_hash", [](const std::vector<at::Tensor>& tensors) {
+    std::vector<XLATensorPtr> xtensors;
+    xtensors.reserve(tensors.size());
+    for (auto& tensor : tensors) {
+      xtensors.push_back(bridge::GetXlaTensor(tensor));
+    }
+    auto hash = XLATensor::GetGraphHash(xtensors);
+    std::string bin((const char*)&hash, sizeof(hash));
+    return py::bytes(bin);
+  });
+
+  m.def("_run_cached_graph",
+        [](const std::string& hash_str,
+           const std::vector<at::IValue>& graph_inputs)
+            -> std::vector<at::Tensor> {
+          TORCH_CHECK(hash_str.size() == sizeof(torch::lazy::hash_t));
+          torch::lazy::hash_t hash = *(torch::lazy::hash_t*)(hash_str.c_str());
+          auto cachedComputation = XLATensor::GetComputationCache()->Get(hash);
+          // TODO implement a fallback mechanism, or make sure those entries
+          // never get kicked out
+          XLA_CHECK(cachedComputation) << "Failed to get computation by hash "
+                                       << torch::lazy::HashToString(hash)
+                                       << ". Maybe the entry get "
+                                          "kicked out of the LRU cache";
+          auto start_prep_input = std::chrono::high_resolution_clock::now();
+
+          // setup the parameters_data
+          std::vector<xla::ComputationClient::DataPtr> parameters_data;
+          auto device = torch_xla::GetCurrentDevice();
```

Comment:
This is OK for now, but I felt like it should be something passed to this function from python. @shunting314 How do you think?